packages/react-reconciler/src/ReactFiberWorkLoop.js

```diff
@@ -2245,8 +2245,9 @@ function renderRootConcurrent(root: FiberRoot, lanes: Lanes) {
             // Selective hydration. An update flowed into a dehydrated tree.
             // Interrupt the current render so the work loop can switch to the
             // hydration lane.
-            workInProgress = null;
             workInProgressRootExitStatus = RootDidNotComplete;
+            // `workInProgress` is not set to `null` to allow unwinding in
+            // `prepareFreshStack`.
```

Comment:
This makes sense but it seems a bit fragile. I think a better approach would be to unwind the stack here in this branch.

You could extract the unwind-y bits from `prepareFreshStack` into a `resetStack` function and call that here.